certora/KilnBase.spec

```diff
@@ -0,0 +1,195 @@
+
+    deny@withrevert(e, usr);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify correct storage changes for not reverting file
+rule file(bytes32 what, uint256 data) {
+    env e;
+
+    uint256 lotBefore = lot();
+    uint256 hopBefore = hop();
+
+    file(e, what, data);
+
+    assert(what == 0x6c6f740000000000000000000000000000000000000000000000000000000000 => lot() == data, "file did not set lot as expected");
+    assert(what != 0x6c6f740000000000000000000000000000000000000000000000000000000000 => lot() == lotBefore, "file did not keep unchanged lot");
+    assert(what == 0x686f700000000000000000000000000000000000000000000000000000000000 => hop() == data, "file did not set hop as expected");
+    assert(what != 0x686f700000000000000000000000000000000000000000000000000000000000 => hop() == hopBefore, "file did not keep unchanged hop");
+}
+
+// Verify revert rules on file
+rule file_revert(bytes32 what, uint256 data) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    file@withrevert(e, what, data);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+    bool revert3 = what != 0x6c6f740000000000000000000000000000000000000000000000000000000000 && // what is not "lot"
+                   what != 0x686f700000000000000000000000000000000000000000000000000000000000;   // what is not "hop"
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+    assert(revert3 => lastReverted, "revert3 failed");
+
+    assert(lastReverted => revert1 || revert2 || revert3, "Revert rules are not covering all the cases");
+}
+
+// Verify rug function on happy path
+rule rug(address dst) {
+    env e;
+
+    require(dai == sell());
+
+    uint256 balanceKilnBefore = dai.balanceOf(currentContract);
+    uint256 balanceDstBefore = dai.balanceOf(dst);
+    uint256 supplyBefore = dai.totalSupply();
+    bool dstSameAsKiln = currentContract == dst;
+
+    rug(e, dst);
+
+    uint256 balanceKilnAfter = dai.balanceOf(currentContract);
+    uint256 balanceDstAfter = dai.balanceOf(dst);
+    uint256 supplyAfter = dai.totalSupply();
+
+    assert(supplyAfter == supplyBefore, "supply did not remain as expected");
+    assert(!dstSameAsKiln => balanceKilnAfter == 0 && balanceKilnBefore == (balanceDstAfter - balanceDstBefore), "balance did not change as expected");
+    assert(dstSameAsKiln => balanceKilnAfter == balanceKilnBefore, "balance changed");
+}
+
+// Verify rug function on reverts
+rule rug_revert(address dst) {
+    env e;
+
+    require(dai == sell());
+
+    uint256 ward = wards(e.msg.sender);
+    uint256 locked = lockedGhost();
+
+    uint256 balanceKilnBefore = dai.balanceOf(currentContract);
```

Comment:
In general for revert rules we do not use "Before" as we only check before as there is not after (reverting). If you still want to use "Before" all the rules should do the same way.

Reply:
yeah, this was just to query the amount of the balance, naming needs to be updated